server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -223,4 +301,25 @@ public String toString() {
     public boolean isEmpty() {
         return shards.isEmpty();
     }
+
+    private boolean invariant() {
+        // initializingShards only contains the initializing shards
+        assert  initializingShards.stream().allMatch(shardRouting -> shardRouting.initializing());
+
+        // relocatingShards only contains the relocating shards
+        assert  relocatingShards.stream().allMatch(shardRouting -> shardRouting.relocating());
+
+        // shards must contains all the shards in initializingShards and relocatingShards
+        assert  initializingShards.stream().allMatch(shardRouting -> shards.containsValue(shardRouting));
```

Comment:
I think we need a stronger property, namely that the order of the shards in `initializingShards` is consistent with that in `shards`:

```suggestion
        assert new ArrayList<>(initializingShards)
            .equals(shards.values().stream().filter(ShardRouting::initializing).collect(Collectors.toList()));
```

(this means the implementation needs some adjustment to get the tests to pass again)